spec/message.abnf

```diff
@@ -9,7 +9,7 @@ selectors = match 1*([s] expression)
 variant = when 1*(s key) [s] pattern
```

Comment:
This is one of two places (with the `option` definition on line 15) where a literal can appear but which cannot be unquoted. I'm wondering if this can introduce confusion for authors? Is this what we want? Here's a couple of examples just to brain-tease. I recognize that `unquoted` is easily confused with `nmtoken` when parsing and that these are probably processed differently, that is, `100` is not equal to `|100|` (the literal is a string, while the nmtoken might be a number). Weakly typed languages might not care, but strongly typed ones might be unhappy.

Is there a way to change our syntax so that _both_ quoted and unquoted literals are always distinct from other types of tokens (such that if I don't have to quote a literal in the `match` or `expression` start I don't have to quote it anywhere)?

```
match horse
when |horse| {requires a quote}
when * {{horse :something option=|horse|} option requires a quote}

match 100
when 100 {100 is nmtoken}
when |100| {100 is literal}
when * {{100 :something option=100} option is nmtoken}
when foo {{100 :something option=|100|} option is literal}
```

Reply:
Just to clarify, an `nmtoken` is valid as a `key`, and `unquoted` is a strict subset of `nmtoken`. Therefore, an `unquoted` token may be used as a `key`.

In my mental model, the MF2 spec level understanding of the type of the value of a `literal`, `nmtoken` or `unquoted` should always be a string. Considering the examples above, a formatting function implementing `:something` should only be passed bags of options like `{ option: 'horse' }` or `{ option: '100' }` irrespective of their presentation in the message syntax as `horse`, `|horse|`, `100`, or `|100|`. Any further parsing of a value as e.g. a number would be entirely the responsibility of the formatting function.

Reply:
The thing that makes me wary is that a parser can "capture" the same value in two different ways--as an `nmtoken` or as an (unquoted) `literal`--suggesting that implementations could produce differing outcomes for the same input (which isn't what we're saying). We should take extra care to describe the relationship between literals and tokens. We don't *say* that `unquoted` is a subset of `nmtoken`--there is neither an ABNF nor specification relationship. 

One challenge is that some functions (such as `plural`) expect _specific_ `nmtoken` values as keys and then separately support a (huge) set of literals, e.g.:

```
match :plural($foo)
when few {an nmtoken recognized by plural}
when 100 {is it an nmtoken or a literal?}
when |-3| {definitely a literal}
when * {etc.}
```

This might be okay, since MF2 is untyped and leaves the parsing up to the functions. But it makes me think we might go back and take up my previous suggestion of merging `unquoted` into the `literal` production. It would be clearer (to me) if (for example) the production for `key` (`nmtoken / literal / "*"`) incorporated `unquoted` via `literal`

Reply:
With the current structure of the ABNF, `unquoted` cannot be a subset of `literal` because that would make the parsing of `key` and `option` ambiguous. For example, with `option=100` we would not be able to say whether the `100` is an `unquoted` via `literal`, or if it's an `nmtoken`.

If we were to make such a change, we could need to limit `nmtoken` so that it _must_ start with either `-` or `:`. This doesn't sound like a good idea to me.

Reply:
> The thing that makes me wary is that a parser can "capture" the same value in two different ways--as an `nmtoken` or as an (unquoted) `literal`--suggesting that implementations could produce differing outcomes for the same input (which isn't what we're saying). We should take extra care to describe the relationship between literals and tokens. We don't _say_ that `unquoted` is a subset of `nmtoken`--there is neither an ABNF nor specification relationship.

We shouldn't think of the ABNF as the definition of the *data model*. Instead, it describes a set of validation rules which can be used to filter and accept the input. As @eemeli says, both `literal` and `unquoted` can end up being encoded in the data model as the same thing. For example, in the implementation that I experimented with, all three: variant keys, expression operands (arguments), and option values are encoded as instances of the [`Literal`](https://github.com/stasm/message2/blob/4abf43f2023b6e20d8ee1d462684d0741ece791b/syntax/ast.ts#L116) class in the data model.

Reply:
> With the current structure of the ABNF, `unquoted` cannot be a subset of `literal` because that would make the parsing of `key` and `option` ambiguous. For example, with `option=100` a parser would not be able to say whether the `100` is an `unquoted` via `literal`, or if it's an `nmtoken`.

Do we need to be able to tell them apart outside of the context in which they appear? We can probably know which is which by inferring it from their position relative to other tokens in the input.

Reply:
I don't think so? That's effectively reflected in #382 as well. But the ABNF can't be ambiguous.

Reply:
@stasm The ABNF is not the definition of the _data model_, but it should, to the extent it is reasonable to do so, be _consistent with_ the data model. I think the thing that is perilous is to have overlap between `literal`, `unquoted`, and `nmtoken` and then describe these separately--with separate requirements--in the data model or elsewhere in the specification. How do I know if `option=100` should apply `unquoted` or `nmtoken` rules to `100`? Is `option=|100|` a different value? Of course, we can avoid that by bringing items together using a production (some productions in the ABNF exist for the convenience of the specification).

Perhaps change this:

```abnf
option = name [s] "=" [s] (literal / nmtoken / variable / unquoted)
```

to something like this:

```abnf
option = name [s] "=" [s] value
value = literal / nmtoken / variable / unquoted
```

(I suggest this for `option` and not for `key` because `key` already does this)